pandas/io/formats/latex.py

```diff
@@ -256,88 +254,394 @@ def _format_multirow(
                     self.clinebuf.append([i + nrow - 1, j + 1])
         return row
 
-    def _print_cline(self, buf: IO[str], i: int, icol: int) -> None:
+    def _compose_cline(self, i: int, icol: int) -> str:
         """
-        Print clines after multirow-blocks are finished.
+        Create clines after multirow-blocks are finished.
         """
+        lst = []
         for cl in self.clinebuf:
             if cl[0] == i:
-                buf.write(f"\\cline{{{cl[1]:d}-{icol:d}}}\n")
-        # remove entries that have been written to buffer
-        self.clinebuf = [x for x in self.clinebuf if x[0] != i]
+                lst.append(f"\n\\cline{{{cl[1]:d}-{icol:d}}}")
+                # remove entries that have been written to buffer
+                self.clinebuf = [x for x in self.clinebuf if x[0] != i]
+        return "".join(lst)
 
-    def _write_tabular_begin(self, buf, column_format: str):
-        """
-        Write the beginning of a tabular environment or
-        nested table/tabular environments including caption and label.
+
+class RowHeaderIterator(RowStringConverter):
+    """Iterator for the table header rows."""
+
+    def __iter__(self):
+        for row_num in range(len(self.strrows)):
+            if row_num < self._header_row_num:
+                yield self.get_strrow(row_num)
+
+
+class RowBodyIterator(RowStringConverter):
+    """Iterator for the table body rows."""
+
+    def __iter__(self):
+        for row_num in range(len(self.strrows)):
+            if row_num >= self._header_row_num:
+                yield self.get_strrow(row_num)
+
+
+class TableBuilderAbstract(ABC):
+    """
+    Abstract table builder producing string representation of LaTeX table.
+    """
+
+    @property
+    def product(self) -> str:
+        """String representation of LaTeX table."""
+        elements = [
+            self.env_begin,
+            self.top_separator,
+            self.header,
+            self.middle_separator,
+            self.env_body,
+            self.bottom_separator,
+            self.env_end,
+        ]
+        result = "\n".join([item for item in elements if item])
+        trailing_newline = "\n"
+        result += trailing_newline
+        return result
+
+    @property
+    @abstractmethod
+    def env_begin(self):
+        """Beginning of the environment."""
+
+    @property
+    @abstractmethod
+    def top_separator(self):
+        """Top level separator."""
+
+    @property
+    @abstractmethod
+    def header(self):
+        """Header lines."""
+
+    @property
+    @abstractmethod
+    def middle_separator(self):
+        """Middle level separator."""
+
+    @property
+    @abstractmethod
+    def env_body(self):
+        """Environment body."""
+
+    @property
+    @abstractmethod
+    def bottom_separator(self):
+        """Bottom level separator."""
+
+    @property
```

Comment:
ok with these being properties, but generaly we like functions if ever an intention of adding args.